### ISSUE TITLE
Question: node grpc versus grpc-js performance

### ISSUE THREAD

**Issue by hardboiled** [grpc/grpc-node#1797]
My team just updated our internal tools to use grpc-js in lieu of grpc. A primary motivation for this is that we want to be able to use later versions of node and no longer use a library that's deprecated.

After a couple weeks, we've identified that the `node 12 + grpc` to `node 14 + grpc-js` upgrade has resulted in significant performance degradation for one of our endpoints. This endpoint makes a protobuf network call for a `repeated` type that returns 100+ objects. Locally we're seeing a 3x latency increase between the two versions of the endpoint, but we haven't debugged the issue more than just swapping the libraries so far, so there is a chance that the issue could be specific to our app.

Is this a known tradeoff between the two libraries, potentially for example that serialization takes longer with the raw JS version? Are there recommended optimizations or configurations that I might not be following that would be contributing to our problem? I would really prefer to keep grpc-js and try to optimize code rather than doing a rollback if possible.

**Comment by murgatroid99** [comment]
Yes, `grpc-js` is known to be slower than the older `grpc`. I don't have any application-level optimizations to suggest to you; I believe the increased latency comes from inefficiencies in the library itself, and/or in the runtime.

**Comment by hardboiled** [comment]
OK, thanks for confirming!

Based on the comment [here](https://github.com/grpc/grpc-node/issues/1588)

> The intended and recommended usage of gRPC is to create a connection once and then use it to make many requests, so in our own benchmarks we generally focus on the request time, and consider the connection establishment time to be amortized over many requests.

We're investigating our connection pooling to see if we can make some improvements here that will mitigate the performance impact. Will report back here with findings.

**Comment by murgatroid99** [comment]
Right, that is a recommendation that applies to both implementations, though the exact performance impact may differ between the two. In general, we do not recommend that you do your own connection pooling. Rather, you should usually create a single `Client` object and use it to make all requests. The exception is if you are running into streams per connection limits on the server, in which case you will need to make multiple clients and force them to use different connections by setting the `grpc.use_local_subchannel_pool` channel option to `1`.

If you want to distribute requests among multiple servers running the same service, you can use the round robin load balancing config to have the client do that, as described in https://github.com/grpc/grpc-node/issues/1307#issuecomment-601312421.

**Comment by hardboiled** [comment]
I spent the last day debugging this, and here are my findings so far... For background, I have a go-based GRPC server running in the cloud, and previously I was using a node-based grpc client also running in the cloud to make requests to this server. 

Upgrading to grpc-js from the legacy grpc resulted in a 4-5x performance degradation on my grpc invocations. Initially I suspected data serialization to be the issue due to our large response bodies, but from running various profilers, it seems like the http2 connections are sitting in async time much longer. I confirmed this theory by running both the node client and the go server locally (without TLS, not sure if that will make a difference here), and each call dropped to a 100ms-ish time.

Upon further inspection, I ran my client with `GRPC_TRACE=call_stream` and `GRPC_VERBOSITY=DEBUG` in both a remote server and local server scenario, and the data frames received from the remote server are much smaller:
```
2021-06-16T10:43:11.965Z | call_stream | [0] receive HTTP/2 data frame of length 9
2021-06-16T10:43:11.965Z | call_stream | [0] receive HTTP/2 data frame of length 4087

**Comment by murgatroid99** [comment]
Can you clarify what you mean by "it seems like the http2 connections are sitting in async time much longer"?

**Comment by hardboiled** [comment]
<img width="738" alt="node-14-16-2-w-grpc-js-library" src="https://user-images.githubusercontent.com/993144/122450617-d3194b80-cf6c-11eb-8113-064da3f1b785.png">
<img width="557" alt="node-14-16-2-w-grpc-legacy-library" src="https://user-images.githubusercontent.com/993144/122450628-d6143c00-cf6c-11eb-81d1-7403c0195789.png">

What I mean is that I wrote a program to make ten grpc calls to the grpc-go server endpoint from the grpc node client and I tested this with the grpc-js library and the grpc legacy library for this node client.

I then used [bubbleprof](https://clinicjs.org/bubbleprof/) to profile the calls, and it's saying that the time waiting on asynchronous calls is ~4x when using the grpc-js library (I included this in the screenshots). Let me know if you need any additional clarification.

**Comment by murgatroid99** [comment]
When it says "async operations were pending", is it talking about time in which no code was running and the Node process was simply sleeping waiting for something to happen, or is it talking about time in which code may or may not be running, but at least one async operation was also pending? Or something else?

In fact, if each request is taking multiple seconds, what makes you think 84 ms of async waiting time is a significant factor? And what is the latency is for running the legacy `grpc` client against the local server? How does that compare to the "100ms+" with grpc-js? Do you still see the same "4-5x performance degradation" as in the other scenario?

That large difference in latency when talking to a local vs remote server makes me think there's something going on here other than just the raw performance of the two libraries. I think it would be helpful to see a full trace log of your main benchmark (grpc-js connecting to a remote server). You can get that by setting the environment variables `GRPC_TRACE=all` and `GRPC_VERBOSITY=DEBUG`.

I do agree that serialization time is likely not the problem, primarily because the serialization code is probably essentially the same in the two tests.

**Comment by hardboiled** [comment]
"is it talking about time in which no code was running and the Node process
was simply sleeping waiting for something to happen, or is it talking about
time in which code may or may not be running, but at least one async
operation was also pending? Or something else?"
I believe it's "sleeping waiting for something to happen" but I can
double-check Monday when i get into the office.

"In fact, if each request is taking multiple seconds, what makes you think
84 ms of async waiting time is a significant factor?"

It's 84,044ms if I'm reading the image i posted right, which is almost the
entire runtime of the program. Are you referring to something else?

"And what is the latency is for running the legacy grpc client against the
local server?"
I haven't run the legacy grpc client yet because the results for the
grpc-js were reasonable. I'm happy to run it Monday if you think that would
be helpful.

"That large difference in latency when talking to a local vs remote server
makes me think there's something going on here other than just the raw
performance of the two libraries."

That's what I'm hoping as well but I've tried to isolate the changes made
to be only the two libraries and I can't explain why the code seems to be
idling for longer. I did compare the output in debug mode yesterday with
the flags you suggested between grpc-js and grpc, but the logging info was
pretty different between the two. I didn't see anything obvious other than
with grpc-js the http2, frames were 1/4 of the size when running against
the remote server as opposed to locally. I can post the results of the
differences between the legacy grpc client and the grpc-js client against
the remote server Monday if you're interested.

**Comment by murgatroid99** [comment]
I guess I misread the numbers in that image. The difference there might just come down to differences in how async operations are represented internally in the two libraries. I don't think you can conclude anything solid from it.

I don't want the differences between the logs. I want to see the grpc-js log. This is about latency; the most important thing is the relative time between all of the different logs.

**Comment by hardboiled** [comment]
Ok, will get the grpc-js logs to you monday then. Thanks for following up!

**Comment by hardboiled** [comment]
Here is the [grpc-js log output](https://gist.github.com/hardboiled/6fbb46193dc87c3ebf66494a29232f50).

**Comment by murgatroid99** [comment]
I assume that is the grpc-js client connecting to the remote server. Can you also share an equivalent log with the grpc-js client connecting to a local server, for comparison?

**Comment by hardboiled** [comment]
Sure, here is the [localhost output](https://gist.github.com/hardboiled/b1976c96cf5e7490f181f12c2f3a7cc3).

**Comment by hardboiled** [comment]
Yeah, the chunk difference was the only thing I noticed between the logs as something that might be some sort of clue. Keep in mind that when running against localhost, I have to use `credentials.createInsecure()` since I don't have valid certs locally, but I figure this detail isn't actually affecting the local performance.

I tried to see if the chunking behavior was different with the legacy library, but I found the logs to be a bit cryptic, so I wasn't able to determine if it's sending more bytes per frame. That being said, when comparing the call-time between the two across 100s of invocations, the difference is 4-5x in latency with a remote server, and this is why we had to rollback our deployment to use the legacy library to avoid production issues.

**Comment by hardboiled** [comment]
```protobuf
message WeK8sGenericService {
  string cluster = 1;
  string namespace = 2;
  string name = 3;
  map<string, string> annotations = 4;
  repeated string hosts = 5;
}

message ListWeK8sGenericServicesResponse {
  repeated WeK8sGenericService services = 1;
}

service Cluster {
  rpc ListWeK8sGenericServices(google.protobuf.Empty) returns (ListWeK8sGenericServicesResponse) {}
}
```

If it helps, this is the message definition.

**Comment by murgatroid99** [comment]
I think what we're seeing there is actually related to TLS, at least partially. When I run my own local large message test with TLS enabled, I see the same pattern of 9 bytes being separated from each chunk into separate chunks. It's hard to tell if that's actually the cause of the performance issues, or if it's incidental.

**Comment by hardboiled** [comment]
Update:
*current solution*:
If you reference my protobuf definitions above, we were adding k8s object annotations (`map<string, string> annotations`) into the `repeated WeK8sGenericService services`. We included these for developer velocity purposes, so that lower-level services wouldn't need to have extra protobuf field definitions updated every time our higher-level graphql server wanted something from the annotations.

When I removed this field and replaced it with one-to-one field definitions for what we actually were pulling, the call time dropped dramatically on local testing to around 100-200ms and all other protobuf calls we benchmarked for grpc-js against grpc legacy were comparable.

*Notes*:
Why did this help? Normally I would think that perhaps the amount of information being returned from the annotations block was just drawfing all else. However, when I outputted the results to a file the file size didn't seem to be proportional to the performance gained from excluding it. I'd have to run more tests to be sure though.

I also tested just using a raw string instead of a map type, and that didn't improve performance at all.

Thus, I'm not sure exactly why this was causing performance problems for my use-case. I would also say that legacy grpc was impacted as well, since it was slow without these changes, just **not as slow** as our grpc-js setup.

Thus, for now, we're happy with just reducing our message size with these changes and leaving it there, even though it's a bit of an inconvenience to have to update field definitions in multiple places every time we want to expose something from annotations to our graphql server.

If you want me to investigate further, I could pull some hard metrics comparing with and without the annotations block for grpc legacy and grpc-js, but for now, I'll regard this issue as resolved. Thanks for your help!

**Comment by murgatroid99** [comment]
I am not surprised that changing the encoding didn't help. The chunking behavior we saw seems more indicative of a problem with how the serialized messages are sent over the wire than a problem with how the messages are serialized. Hopefully that means that we can reproduce the problem by sending any large messages over the network, instead of relying on your specific protocol.

**Comment by bobaaaaa** [comment]
I guess we run into the same problem. We downgrade to the (deprecated) `grpc` package. That helps, but is unsatisfying. I could also provide trace logs if that helps.

**Comment by edvardchen** [comment]
grpc-js uses built-in http module and node grpc uses their own one. That means grpc-js can leverage multiple cpu core in Node.js cluster mode but node-grpc can't

**Comment by Farenheith** [comment]
I also had performance related problems when I tried to upgrade from grpc to grpc-js, but what I've seen is the application almost doubling the number of task when using the new library. I just downgraded to grpc, and the number of tasks came back the the previous one, both tests were on node 14. The autoscaling here is configured by CPU usage, so, the conclusion is that grpc-js demands much more cpu than grpc. Could it be possible that this will not be the case if I also upgrade to the latest node version?

**Comment by kirillgroshkov** [comment]
> almost doubling the number of task when using the new library

Just to add to that - the speed of streaming the query from Datastore to Node halved with the update from `grpc` to `grpc-js` (e.g from 2000 rows per second down to 1000 rows per second). To keep in mind.

**Comment by igorescobar-moonpay** [comment]
It seems that, for our use case, we are also stuck with `grpc` since `grpc-js` is significantly slower for our use case. We just can't keep up with the data flow using `grpc-js` putting our data ingestion stale. We are connected to a streaming that are pushing ~10-15mb of data every 400ms. No issues with `grpc` and  with `grpc-js` the data starts to get more and more stale over time.

**Comment by Farenheith** [comment]
> It seems that, for our use case, we are also stuck with `grpc` since `grpc-js` is significantly slower for our use case. We just can't keep up with the data flow using `grpc-js` putting our data ingestion stale. We are connected to a streaming that are pushing ~10-15mb of data every 400ms. No issues with `grpc` and  with `grpc-js` the data starts to get more and more stale over time.

I wonder if it would be possible to fork grpc and create a version compatible with latest node. E company I'm working at right now if not using grpc at all, but it seems like a interesting project to look on
